### PR TITLE
chore: Refactor Python Codebase

### PR DESCRIPTION
**Pull request by koconder**
Work in progress,

Some refactoring work, will provide change log and tests after completion.

## Checklist
- [x] The PR conforms to DataHub's [Contributing Guideline](https://github.com/datahub-project/datahub/blob/master/docs/CONTRIBUTING.md) (particularly [Commit Message Format](https://github.com/datahub-project/datahub/blob/master/docs/CONTRIBUTING.md#commit-message-format))
- [x] Links to related issues (if applicable)
- [x] Tests for the changes have been added/updated (if applicable)
- [x] Docs related to the changes have been added/updated (if applicable). If a new feature has been added a Usage Guide has been added for the same.
- [x] For any breaking change/potential downtime/deprecation/big changes an entry has been made in [Updating DataHub](https://github.com/datahub-project/datahub/blob/master/docs/how/updating-datahub.md)

### CODE REVIEW

**Comment by koconder** [comment]
`tests/integration/hana/test_hana.py::test_hana_ingest FAILED` but no changes to any codebase related to hana ingestion

File: metadata-ingestion/examples/bootstrap_data/business_glossary.yml [hunk 1]
```diff
@@ -40,7 +40,7 @@ nodes:
         inherits:
           - Classification.Sensitive
   - name: ClientsAndAccounts
-    description: Provides basic concepts such as account, account holder, account provider, relationship manager that are commonly used by financial services providers to describe customers and to determine counterparty identities
```
Comment (anshbansal) [inline metadata-ingestion/examples/bootstrap_data/business_glossary.yml:43]: `counterparty` is correct here. Please revert this change.

**Comment by koconder** [comment]
I will re-run the linter as i can see some changes are needed by some of the tests

**Comment by koconder** [comment]
mypy raising the following issue which im unable to trace - Not sure if someone can help @maggiehays 
```
src/datahub/cli/cli_utils.py:582: error: Incompatible return value type (got "Union[int, Any]", expected "Dict[Any, Any]")
```

**Comment by koconder** [comment]
Ready, mypy tests now pass locally

**Comment by koconder** [comment]
@anshbansal checks hava passed ready to merge. Once merged ill raise another PR for pre-commit hooks and lint rules

File: metadata-ingestion/src/datahub/ingestion/source/looker.py [hunk 2]
```diff
@@ -449,12 +447,8 @@ def _get_looker_dashboard_element(  # noqa: C901
             raise ValueError("Element ID can't be None")
 
         if element.query is not None:
-            explores = []
             fields = self._get_fields_from_query(element.query)
-            if element.query.view is not None:
-                # Get the explore from the view directly
```
Comment (anshbansal) [inline metadata-ingestion/src/datahub/ingestion/source/looker.py:455]: Please add back this comment.

File: metadata-ingestion/src/datahub/ingestion/source/looker.py [hunk 3]
```diff
@@ -512,7 +505,6 @@ def _get_looker_dashboard_element(  # noqa: C901
                     upstream_fields=fields,
                 )
 
-        # Failing the above two approaches, pick out details from result_maker
```
Comment (anshbansal) [inline metadata-ingestion/src/datahub/ingestion/source/looker.py:515]: Please add back this comment

File: metadata-ingestion/src/datahub/ingestion/source/lookml.py [hunk 4]
```diff
@@ -654,14 +651,7 @@ def from_looker_dict(
...
         # object in the database.
-        if sql_table_name is not None:
-            # If sql_table_name is set, there is a single dependency in the view, on the sql_table_name.
-            sql_table_names = [sql_table_name]
-        else:
-            # Otherwise, default to the view name as per the docs:
-            # https://docs.looker.com/reference/view-params/sql_table_name-for-view
-            sql_table_names = [view_name]
-
+        sql_table_names = [view_name] if sql_table_name is None else [sql_table_name]
```
Comment (anshbansal) [inline metadata-ingestion/src/datahub/ingestion/source/lookml.py:654]: Please revert this change or add back the comments.

File: metadata-ingestion/src/datahub/ingestion/source_config/pulsar.py [hunk 5]
```diff
@@ -30,7 +30,7 @@ def _is_valid_hostname(hostname: str) -> bool:
     """
     if len(hostname) > 253:
         return False
-    # Hostnames ending on a dot are valid, if present strip exactly one
```
Comment (anshbansal) [inline metadata-ingestion/src/datahub/ingestion/source_config/pulsar.py:33]: Please revert this change.

File: metadata-ingestion/src/datahub/integrations/great_expectations/action.py [hunk 6]
```diff
@@ -678,9 +675,9 @@ def get_dataset_partitions(self, batch_identifier, data_asset):
         return dataset_partitions
 
     def get_platform_instance(self, datasource_name):
-        if self.platform_instance_map and datasource_name in self.platform_instance_map:
-            return self.platform_instance_map[datasource_name]
         if self.platform_instance_map:
+            if datasource_name in self.platform_instance_map:
```
Comment (anshbansal) [inline metadata-ingestion/src/datahub/integrations/great_expectations/action.py:679]: Please explain this change.

File: metadata-ingestion/src/datahub/configuration/common.py [hunk 7]
```diff
@@ -181,28 +173,21 @@ def all(cls) -> "KeyValuePattern":
...
     def matched(self, string: str) -> bool:
-        for key in self.rules.keys():
-            if re.match(key, string):
-                return True
-        return False
+        return any(re.match(key, string) for key in self.rules.keys())
 
     def is_fully_specified_key(self) -> bool:
         """
-        If the allow patterns are literals and not full regexes, then it is considered
```
Comment (anshbansal) [inline metadata-ingestion/src/datahub/configuration/common.py:197]: please revert

**Review by anshbansal — CHANGES_REQUESTED** [review]

File: metadata-ingestion/src/datahub/configuration/common.py [hunk 8]
```diff
@@ -104,7 +101,7 @@ class OauthConfiguration(ConfigModel):
 
 
 class AllowDenyPattern(ConfigModel):
-    """A class to store allow deny regexes"""
```
Comment (anshbansal) [inline metadata-ingestion/src/datahub/configuration/common.py:107]: The plural `regexes` was correct. It is not `regexs`. Please change back in all places

**Review by anshbansal — CHANGES_REQUESTED** [review]
One change was not done completely. Please do that. Then we can merge this in.

**Comment by koconder** [comment]
@anshbansal i have done the missing typo change, and i have updated the linting so checks should all pass

**Comment by anshbansal** [comment]
Linting is still failing
```
+ mypy src/ tests/ examples/
0
src/datahub/utilities/mapping.py:235: error: Incompatible return value type (got "bool", expected "Optional[Match[Any]]")
src/datahub/utilities/mapping.py:237: error: Incompatible return value type (got "bool", expected "Optional[Match[Any]]")
Found 2 errors in 1 file (checked 479 source files)

```

**Comment by anshbansal** [comment]
@koconder Request you to review all of the code yourself once and ensure you are not changing the contracts of any methods. This is supposed to be a refactor only. No functionality changes please. If you wish to make any functionality changes that should be a separate Pull Request. Request you to go through https://datahubproject.io/docs/metadata-ingestion/developing/#contributing and ensure linting passes. This will ensure we can get this PR merged.

**Comment by koconder** [comment]
@anshbansal sorry about that, all checks passing locally. All changes requested made

**Comment by koconder** [comment]
@anshbansal all my code is fine, however some changes from two months ago playing up now? I'll see if i can fix the conflict/issue but seems like was introduced by someone else. I keep having to play catchup on other peoples code

```
src/datahub/ingestion/source/s3/source.py:597:16: F821 undefined name 'S3BucketKey'
src/datahub/ingestion/source/s3/source.py:607:19: F821 undefined name 'GlobalTagsClass'
src/datahub/ingestion/source/s3/source.py:610:20: F821 undefined name 'GlobalTagsClass'
src/datahub/ingestion/source/s3/source.py:618:25: F821 undefined name 'make_tag_urn'
src/datahub/ingestion/source/s3/source.py:634:25: F821 undefined name 'make_tag_urn'
src/datahub/ingestion/source/s3/source.py:646:36: F821 undefined name 'GlobalTagsClass'
src/datahub/ingestion/source/s3/source.py:649:29: F821 undefined name 'GlobalTagsClass'
src/datahub/ingestion/source/s3/source.py:661:20: F821 undefined name 'GlobalTagsClass'
src/datahub/ingestion/source/s3/source.py:662:19: F821 undefined name 'TagAssociationClass'
src/datahub/ingestion/source/s3/source.py:667:16: F821 undefined name 'FolderKey'
```

**Comment by koconder** [comment]
@anshbansal merge conflicts all sorted now, ready to merge. I have run all the tests again locally

File: metadata-ingestion/src/datahub/utilities/mapping.py [hunk 9]
```diff
@@ -232,8 +232,8 @@ def get_match(self, match_clause: Any, raw_props_value: Any) -> Optional[Match]:
         if type(raw_props_value) not in Constants.OPERAND_DATATYPE_SUPPORTED or type(
             raw_props_value
         ) != type(match_clause):
-            return None
+            return False
```
Comment (anshbansal) [inline metadata-ingestion/src/datahub/utilities/mapping.py:235]: This is returning `bool` instead of `Optional[Match]`. This changes the contract of this method. You marked my earlier comment as resolved. Please explain why are you changing the contract here?

**Review by anshbansal — CHANGES_REQUESTED** [review]
Please do not change the contract. Please revert any change in contracts you have done in this PR.

**Review by anshbansal — APPROVED** [review]
LGTM will wait for tests to pass
